pkg/kubelet/pod_workers.go

```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+	// enforce the restriction that a grace period can only decrease and track whatever our value is,
+	// then ensure a calculated value is passed down to lower levels
+	gracePeriod := status.gracePeriod
+	// this value is bedrock truth - the apiserver owns telling us this value calculated by apiserver
+	if override := pod.DeletionGracePeriodSeconds; override != nil {
+		if gracePeriod == 0 || *override < gracePeriod {
+			gracePeriod = *override
+		}
+	}
+	// we allow other parts of the kubelet (namely eviction) to request this pod be terminated faster
+	if options != nil {
+		if override := options.PodTerminationGracePeriodSecondsOverride; override != nil {
+			if gracePeriod == 0 || *override < gracePeriod {
+				gracePeriod = *override
+			}
+		}
+	}
+	// make a best effort to default this value to the pod's desired intent, in the event
+	// the kubelet provided no requested value (graceful termination?)
+	if gracePeriod == 0 && pod.Spec.TerminationGracePeriodSeconds != nil {
+		gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
+	}
+	// no matter what, we always supply a grace period of 1
+	if gracePeriod < 1 {
+		gracePeriod = 1
+	}
+	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
+}
+
+func (p *podWorkers) managePodLoop(podUpdates <-chan UpdatePodOptions) {
+	var lastSyncTime time.Time
+	var lastKillTime time.Time
+	for update := range podUpdates {
+		klog.V(4).InfoS("DEBUG: Processing pod event", "pod", klog.KObj(update.Pod), "podUID", update.Pod.UID, "updateType", update.UpdateType)
+		err := func() error {
+			// The worker is responsible for ensuring the sync method sees the appropriate
+			// status updates on resyncs (the result of the last sync), transitions to
+			// terminating (no wait), or on terminated (whatever the most recent state is).
+			// Only syncing and terminating can generate pod status changes, while terminated
+			// pods ensure the most recent status makes it to the api server.
+			var status *kubecontainer.PodStatus
+			var err error
+			switch {
+			case update.RunningPod != nil:
+				// when we receive a running pod, we don't need status at all to terminate a pod
+			case update.UpdateType == kubetypes.SyncPodKill:
+				if lastKillTime.IsZero() {
+					// if this is the first time we are killing the pod, we don't want to wait on the pod
+					// cache at all as it may be out of date and we want to begin termination as quickly
+					// as possible
+					lastKillTime = time.Now()
+					status = nil
+					break
+				}
+				if lastSyncTime.IsZero() {
+					// in the event we have never synced this pod before, we may have no container status
+					status = nil
+					break
+				}
+				// if this is a retry of termination or a delete shorten, we expect cache to eventually get
+				// filled and can thus simply use the last syncTime
+				status, err = p.podCache.GetNewerThan(update.Pod.UID, lastSyncTime)
```

Comment:
If we are OK with waiting for the new state to be populated, we should use `GetNewerThan` here, we can even use `time.Now()` as the minTime.

If we don't want to wait for it for some reason, we should call an `UpdateCache` here.

Passing `status = nil` makes it a bit hard to understand the code.

Reply:
I was uncomfortable with having podWorker update the status cache - that's sync*Pods job right now (to refresh).  I didn't feel like the logic for syncTerminating/TerminatedPod was more readable by having status refresh outside of their loops.  Agree status == nil was wierd, but I think it might actually possible to get into syncTerminatingPod and status manager has never gotten status.  

We are not ok with waiting for updated state here, because it's possible that you delete a pod BEFORE syncPod gets far enough to create any containers (especially when context cancellation goes in), and therefore you would have to wait for the next pleg refresh (thus increasing the amount of time a create then delete takes).  In general I think we should bias towards lower latency on delete/kill/evict, so I didn't want to wait in those cases.

I considered passing the defaulted Get status, but wanted to distinguish the case where we definitely want to refresh the state of the pod before delete (and syncTerminatingPod felt like the right place to handle it).  Let me mull on this a bit though if that's not sufficiently convincing.